bundles/org.eclipse.jface/src/org/eclipse/jface/resource/ResourceManager.java

```diff
@@ -102,51 +102,172 @@ public abstract class ResourceManager {
+	 * the descriptor. For example, when given an ImageDescriptor, the return value
+	 * of this method may be downcast to Image. If possible it is recommended to use
+	 * the more specific overloads of this methods, that returned the corresponding
+	 * resource sub-class.
 	 * </p>
 	 *
 	 * <p>
-	 * This method should only be used for resources that should remain
-	 * allocated for the lifetime of the resource manager. To allocate shorter-lived
+	 * This method should only be used for resources that should remain allocated
+	 * for the lifetime of this resource-manager. To allocate shorter-lived
 	 * resources, manage them with <code>create</code>, and <code>destroy</code>
 	 * rather than this method.
 	 * </p>
 	 *
 	 * <p>
-	 * This method should never be called on the global resource manager,
-	 * since all resources will remain allocated for the lifetime of the app and
-	 * will be effectively leaked.
+	 * This method should never be called on the global resource-manager, since all
+	 * resources will remain allocated for the lifetime of the app and will be
+	 * effectively leaked.
 	 * </p>
 	 *
 	 * @param descriptor identifier for the requested resource
-	 * @return the requested resource. Never null.
+	 * @return the requested resource, never null
 	 * @throws DeviceResourceException if the resource does not exist yet and cannot
-	 * be created for any reason.
+	 *                                 be created for any reason.
 	 *
 	 * @since 3.3
+	 * @see ResourceManager#get(ImageDescriptor)
+	 * @see ResourceManager#get(ColorDescriptor)
+	 * @see ResourceManager#get(FontDescriptor)
 	 */
 	public final Object get(DeviceResourceDescriptor descriptor) {
-		Object result = find(descriptor);
+		return getCachedResource(descriptor);
+	}
 
-		if (result == null) {
-			result = create(descriptor);
-		}
+	/**
+	 * Returns a previously-allocated {@link Image} or allocates a new one if none
+	 * exists yet. The image will remain allocated for at least the lifetime of this
+	 * resource-manager. If necessary, the image will be deallocated automatically
+	 * when this resource-manager is disposed.
+	 * <p>
+	 * The images returned from this method are reference counted and may be shared
+	 * internally with other resource-managers. They should never be disposed
+	 * outside of the ResourceManager framework, or it will cause exceptions in
+	 * other code that shares them. For example, never call
+	 * {@link org.eclipse.swt.graphics.Image#dispose()} on anything returned from
+	 * this method.
+	 * </p>
+	 *
+	 * <p>
+	 * This method should only be used for images that should remain allocated for
+	 * the lifetime of this resource-manager. To allocate shorter-lived images,
+	 * manage them with {@link #createImage(ImageDescriptor)}, and
+	 * {@link #destroyImage(ImageDescriptor)} rather than this method.
+	 * </p>
+	 *
+	 * <p>
+	 * This method should never be called on the global resource-manager, since all
```

Comment:
I don't like a method that should not be called. Better use a specialized interface for Managers that "support" the method.

Reply:
This method already exists... But we should annotate it with @Deprecated if it should not be called (anymore).

Reply:
> I don't like a method that should not be called. Better use a specialized interface for Managers that "support" the method.

I understand your point, but it should only not be called for the global resource-manager (e.g. returned by `JFaceResources.getResources()`) for all others it is fine. I don't have a direct use-case at hand, but even for a global manager it could make sense to call this method if one really wants a resource for the entire lifetime of the application (of course in general this would be a bad practice).

That being said, alternatively those new methods could be added to the `LocalResourceManager`. Then at least the new methods one cannot call (accidentally) on a global `DeviceResourceManager`.
But only adding it to `LocalResourceManager` would probably hinder the adoption a bit, because AFAICT usually one uses the `ResourceManager` type for variables and then does not stumble about this more strictly typed methods.

And given that there is already a `get` method, I tend to perfer to add them at the same location. If all get's were new, then yes, should they have been placed at `LocalResourceManager`.

> This method already exists... But we should annotate it with @deprecated if it should not be called (anymore).

For cases where one has a custom sub-class of `DeviceResourceDescriptor` that does handle something else like images, colors or font this method could still be relevant. And the Java-Doc of `DeviceResourceDescriptor` does not forbid direct subclassing.

Reply:
then the javadoc never call on "global resource-manager" is just wrong.